3scaleAdapter/pkg/threescale/metrics/prometheus.go

```diff
@@ -65,18 +104,41 @@ func NewMetricsReporter(reportMetrics bool, serveOnPort int) *Reporter {
 	return &Reporter{reportMetrics, serveOnPort}
 }
 
-// ObserveSystemLatency reports a metric to system latency histogram
-func (r *Reporter) ObserveSystemLatency(sysURL string, serviceID string, observed time.Duration) {
+// ObserveLatency reports a metric to a latency histogram
+func (r *Reporter) ObserveLatency(serviceID string, l LatencyReport) {
 	if r != nil && r.shouldReport {
-		systemLatency.WithLabelValues(sysURL, serviceID).Observe(observed.Seconds())
+		if l.URL != "" {
+			switch l.Target {
```

Comment:
perhaps we can move this switch to a method in a `LatencyReport` object that generates a `prometheus.Observer` to be used here so that we just call `o := l.GetObserver(serviceID)` and then `o.Observe(l.TimeTaken)` (unless the `r.observeLatency` helper is useful elsewhere). We could maybe even split this into smaller helpers, but this would be a good first step to make this method smaller and clearer.